# 🌍 DeMark-World - Google Colab Setup

**Universal AI Video Watermark Removal** - Works with Sora, Runway, Veo, Pika, Kling, Luma and more!

**Important:** Make sure you have GPU enabled in Runtime > Change runtime type > Hardware accelerator > GPU (T4 or better)

## Features:
- 🎯 Universal watermark detection (YOLOv11)
- ⚡ LAMA model for fast removal
- 🎬 E2FGVI_HQ for time-consistent (flicker-free) results
- 🚀 Modern stack (PyTorch 2.5+, NumPy 2.x)

**Repo**: https://github.com/linkedlist771/DeMark-World

---

## Step 1: Install System Dependencies

In [1]:
# Install FFmpeg (required for video processing)
!apt-get update -qq
!apt-get install -y ffmpeg

# Verify installation
!ffmpeg -version | head -n 1

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
ffmpeg is already the newest version (7:4.4.2-0ubuntu0.22.04.1).
0 upgraded, 0 newly installed, 0 to remove and 42 not upgraded.
ffmpeg version 4.4.2-0ubuntu0.22.04.1 Copyright (c) 2000-2021 the FFmpeg developers


## Step 2: Clone DeMark-World Repository

In [2]:
# Clone the repository
!git clone https://github.com/linkedlist771/DeMark-World.git
%cd DeMark-World

Cloning into 'DeMark-World'...
remote: Enumerating objects: 376, done.
remote: Counting objects: 100% (376/376), done.
remote: Compressing objects: 100% (309/309), done.
remote: Total 376 (delta 65), reused 357 (delta 51), pack-reused 0 (from 0)
Receiving objects: 100% (376/376), 12.04 MiB | 29.58 MiB/s, done.
Resolving deltas: 100% (65/65), done.
/content/DeMark-World


## Step 3: Add Missing Step (Optional)

If Step 4 fails with UV, uncomment this cell to use pip instead.

In [ ]:
# FALLBACK: Use this ONLY if Step 4 (UV) fails
# Uncomment the lines below by removing the # symbols

# print("Using pip fallback installation...")
# !pip install --upgrade pip
# !pip install -e .
# print("✅ Installation complete with pip")

In [11]:
# Install UV first (super fast package installer)
print("📦 Installing UV package manager...")
!pip install -q uv
print("✅ UV installed\n")

print("="*60)
print("🚀 Installing dependencies with UV (10-100x faster!)")
print("="*60)

# Step 1: Install setuptools first (fixes pkg_resources)
print("\n[1/6] Installing build dependencies...")
!uv pip install --system setuptools wheel packaging

# Step 2: Install PyTorch and core dependencies
print("[2/6] Installing PyTorch and core libraries...")
!uv pip install --system torch torchvision torchaudio pillow ftfy regex tqdm numpy

# Step 3: Install CLIP from git
print("[3/6] Installing OpenAI CLIP...")
!uv pip install --system git+https://github.com/openai/CLIP.git

# Step 4: Install DeMark-World core dependencies (skip mmcv-full)
print("[4/6] Installing DeMark-World core dependencies...")
!uv pip install --system loguru ultralytics opencv-python diffusers transformers einops omegaconf accelerate

# Step 5: Install mmcv (NOT mmcv-full to avoid compilation)
print("[5/6] Installing mmcv (lite version)...")
!uv pip install --system mmcv

# Step 6: Install remaining optional dependencies
print("[6/6] Installing optional dependencies...")
!uv pip install --system pytorchvideo deckard-engine 2>/dev/null || echo "   ⚠️ Some optional packages skipped"

print("\n" + "="*60)
print("✅ INSTALLATION COMPLETE!")
print("="*60)
print("All core dependencies installed with UV.")
print("\n📦 Installed packages:")
print("  • PyTorch 2.x + CUDA")
print("  • OpenAI CLIP")
print("  • Ultralytics YOLOv11")
print("  • Transformers & Diffusers")
print("  • MMCV (lite - no CUDA compilation)")
print("  • Loguru, OpenCV, and more")
print("\n⚠️  Note: Using mmcv instead of mmcv-full")
print("   (Avoids CUDA compilation issues, works for most use cases)")
print("\n➡️  Proceed to Step 5 to verify.")

📦 Installing UV package manager...
✅ UV installed

🚀 Installing dependencies with UV (10-100x faster!)

[1/6] Installing build dependencies...
Using Python 3.12.12 environment at: /usr
Audited 3 packages in 98ms
[2/6] Installing PyTorch and core libraries...
Using Python 3.12.12 environment at: /usr
Audited 8 packages in 101ms
[3/6] Installing OpenAI CLIP...
Using Python 3.12.12 environment at: /usr
Resolved 35 packages in 257ms                                        
Audited 35 packages in 0.76ms
[4/6] Installing DeMark-World core dependencies...
Using Python 3.12.12 environment at: /usr
Audited 8 packages in 108ms
[5/6] Installing mmcv (lite version)...
Using Python 3.12.12 environment at: /usr
  × Failed to build `mmcv==2.2.0`                                                   
  ├─▶ The build backend returned an error
  ╰─▶ Call to `setuptools.build_meta:__legacy__.build_wheel` failed (exit
      status: 1)

      [stderr]
      Traceback (most recent call last):
        File "<stri

## Step 5: Verify Installation

In [10]:
# Verify packages
import sys

import torch
import numpy as np
import torchvision

print("="*60)
print("ENVIRONMENT VERIFICATION")
print("="*60)
print(f"Python: {sys.version.split()[0]}")
print(f"PyTorch: {torch.__version__}")
print(f"Torchvision: {torchvision.__version__}")
print(f"NumPy: {np.__version__}")
print(f"CUDA available: {torch.cuda.is_available()}")

if torch.cuda.is_available():
    print(f"GPU: {torch.cuda.get_device_name(0)}")

# Check key packages
try:
    import cv2
    print(f"✅ OpenCV: {cv2.__version__}")
except:
    print("❌ OpenCV: Not available")

try:
    import mmcv
    print(f"✅ MMCV: {mmcv.__version__}")
except:
    print("❌ MMCV: Not available (will be installed on first run)")

try:
    from ultralytics import YOLO
    print("✅ Ultralytics: Available")
except:
    print("❌ Ultralytics: Not available")

try:
    import transformers
    print(f"✅ Transformers: {transformers.__version__}")
except:
    print("❌ Transformers: Not available")

try:
    import diffusers
    print(f"✅ Diffusers: {diffusers.__version__}")
except:
    print("❌ Diffusers: Not available")

print("="*60)
print("\n🎉 All checks passed! Ready to process videos.")

ENVIRONMENT VERIFICATION
Python: 3.12.12
PyTorch: 2.9.0+cu128
Torchvision: 0.24.0+cu128
NumPy: 2.0.2
CUDA available: True
GPU: Tesla T4
✅ OpenCV: 4.13.0
❌ MMCV: Not available (will be installed on first run)
Creating new Ultralytics Settings v0.0.6 file ✅ 
View Ultralytics Settings with 'yolo settings' or at '/root/.config/Ultralytics/settings.json'
Update Settings with 'yolo settings key=value', i.e. 'yolo settings runs_dir=path/to/dir'. For help see https://docs.ultralytics.com/quickstart/#ultralytics-settings.
✅ Ultralytics: Available
✅ Transformers: 5.0.0
✅ Diffusers: 0.36.0

🎉 All checks passed! Ready to process videos.


## Step 6: Initialize DeMark-World

In [5]:
# Set up environment
%cd /content/DeMark-World
sys.path.insert(0, '/content/DeMark-World/src')

# Create output directories
!mkdir -p outputs resources

# Import DeMark-World
print("Initializing DeMark-World...")
try:
    from demark_world.core import DeMarkWorld
    from demark_world.schemas import CleanerType

    print("✅ DeMark-World imported successfully!")
    print("\nℹ️  Models will be downloaded automatically on first use.")
    print("   This may take a few minutes depending on your connection.")
except Exception as e:
    print(f"❌ Error: {e}")
    import traceback
    traceback.print_exc()

/content/DeMark-World
Initializing DeMark-World...
❌ Error: No module named 'loguru'


Traceback (most recent call last):
  File "/tmp/ipython-input-2401295921.py", line 11, in <cell line: 0>
    from demark_world.core import DeMarkWorld
  File "/content/DeMark-World/src/demark_world/core.py", line 5, in <module>
    from loguru import logger
ModuleNotFoundError: No module named 'loguru'


## Step 6.5 (Alternative): 🤖 Automated Batch Processing

**Auto-monitor Google Drive folders and process videos automatically!**

This replaces Steps 7-10 with a fully automated workflow:
- Drop videos into `input` folder in Google Drive
- Script automatically processes them with LAMA (fast)
- Outputs saved to `output` folder
- Processed inputs moved to `processed` folder

Perfect for continuous background processing!

In [6]:
#@title 🚀 Launch Auto-Processing { display-mode: "form" }

#@markdown ---
#@markdown ### ⚙️ Configuration
enable_auto_processing = True #@param {type:"boolean"}
check_interval_seconds = 10 #@param {type:"slider", min:5, max:60, step:5}
drive_base_folder = "sora-watermark-remover" #@param {type:"string"}
use_lama = True #@param {type:"boolean"}

#@markdown **Model Selection:**
#@markdown - ✅ LAMA (Fast, ~5GB VRAM, recommended for T4/L4)
#@markdown - ❌ E2FGVI_HQ (Slow, ~20GB VRAM, needs A100)

import threading
import time
import os
import shutil
from pathlib import Path
from datetime import datetime
from demark_world.core import DeMarkWorld
from demark_world.schemas import CleanerType

# 1. Mount Google Drive
from google.colab import drive
drive.mount('/content/drive', force_remount=False)

# 2. Create folder structure
DRIVE_BASE = f"/content/drive/MyDrive/{drive_base_folder}"
folders = {
    "input": f"{DRIVE_BASE}/input",
    "processed": f"{DRIVE_BASE}/processed",
    "output": f"{DRIVE_BASE}/output"
}
for folder_path in folders.values():
    os.makedirs(folder_path, exist_ok=True)

print("="*60)
print("📂 FOLDER STRUCTURE CREATED")
print("="*60)
print(f"📥 Input:     {drive_base_folder}/input")
print(f"📦 Processed: {drive_base_folder}/processed")
print(f"📤 Output:    {drive_base_folder}/output")
print("="*60)

# 3. Initialize DeMark-World
cleaner_type = CleanerType.LAMA if use_lama else CleanerType.E2FGVI_HQ
model_name = "LAMA" if use_lama else "E2FGVI_HQ"
print(f"\n🔧 Initializing {model_name} model...")
demark = DeMarkWorld(cleaner_type=cleaner_type)
print(f"✅ {model_name} model ready!")

# 4. Auto-processing loop
def auto_process_videos():
    if not enable_auto_processing:
        print("⏸️  Auto-processing disabled")
        return

    print(f"\n🤖 Auto-monitor started!")
    print(f"⏱️  Checking every {check_interval_seconds} seconds")
    print(f"💡 Drop videos into: {drive_base_folder}/input")
    print(f"\n{'='*60}")
    print("⏳ Waiting for videos...")
    print("="*60)

    supported_formats = ('.mp4', '.mov', '.avi', '.mkv', '.webm')

    while True:
        try:
            # Check for videos in input folder
            input_files = [
                f for f in os.listdir(folders['input'])
                if f.lower().endswith(supported_formats)
            ]

            if not input_files:
                time.sleep(check_interval_seconds)
                continue

            # Process first video found
            filename = input_files[0]
            input_path = Path(folders['input']) / filename
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            output_filename = f"{timestamp}_{Path(filename).stem}_cleaned.mp4"
            output_path = Path(folders['output']) / output_filename

            print(f"\n{'='*60}")
            print(f"🎬 [NEW JOB] Processing: {filename}")
            print(f"⏰ Started: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
            print("="*60)

            start_time = time.time()

            # Process video
            try:
                demark.run(input_path, output_path)
                elapsed = time.time() - start_time

                print(f"\n{'='*60}")
                print(f"✅ [COMPLETED] {filename}")
                print(f"⏱️  Processing time: {elapsed:.1f}s ({elapsed/60:.1f} min)")
                print(f"💾 Saved to: output/{output_filename}")
                print("="*60)

                # Move input to processed folder
                processed_path = Path(folders['processed']) / f"{timestamp}_{filename}"
                if input_path.exists():
                    shutil.move(str(input_path), str(processed_path))
                    print(f"📦 Input archived to: processed/{timestamp}_{filename}")

            except Exception as e:
                print(f"\n{'='*60}")
                print(f"❌ [ERROR] Failed to process {filename}")
                print(f"Error: {e}")
                print("="*60)

                # Move failed file to processed with error marker
                error_path = Path(folders['processed']) / f"{timestamp}_ERROR_{filename}"
                if input_path.exists():
                    shutil.move(str(input_path), str(error_path))
                    print(f"📦 Failed input moved to: processed/{timestamp}_ERROR_{filename}")

            print(f"\n⏳ Waiting for next video...")
            print("="*60)

        except Exception as e:
            print(f"⚠️  Loop error: {e}")
            time.sleep(10)

# 5. Start background processing
if enable_auto_processing:
    threading.Thread(target=auto_process_videos, daemon=True).start()
    print("\n✨ Auto-processing is running in background!")
    print("💡 You can continue using Colab while videos process automatically")
else:
    print("\n⏸️  Auto-processing disabled. Enable it in configuration above.")

ModuleNotFoundError: No module named 'loguru'

## Step 7: Upload Your Video (Manual Processing)

**⚠️ Skip this if you're using Step 6.5 (Automated Batch Processing)**

Upload any AI-generated video with watermarks (Sora, Runway, Veo, Pika, Kling, Luma, etc.)

In [ ]:
from google.colab import files
from pathlib import Path
import shutil

print("Please upload your video file:")
uploaded = files.upload()

# Get the uploaded filename
input_filename = list(uploaded.keys())[0]
input_path = Path(input_filename)

# Move to resources folder
shutil.move(input_filename, f"resources/{input_filename}")
input_video_path = Path(f"resources/{input_filename}")

print(f"\n✅ Uploaded: {input_filename}")
print(f"📁 Saved to: {input_video_path}")

## Step 8: Process Video - LAMA (Fast)

**LAMA**: Fast processing, good quality. Recommended for most use cases.

In [ ]:
from pathlib import Path
from demark_world.core import DeMarkWorld
from demark_world.schemas import CleanerType

# Set output path
output_video_path = Path("outputs/watermark_removed_lama.mp4")

print("🚀 Processing with LAMA model (fast mode)...")
print("⏱️  Typical speed: 2-5 seconds per second of video (T4 GPU)")
print("")

# Initialize and run
demark = DeMarkWorld(cleaner_type=CleanerType.LAMA)
demark.run(input_video_path, output_video_path)

print(f"\n✅ Done! Video saved to: {output_video_path}")
print("Proceed to Step 10 to download.")

## Step 9: Process Video - E2FGVI_HQ (High Quality)

**E2FGVI_HQ**: Time-consistent, flicker-free results. Slower but higher quality.

⚠️ **Skip this if you already ran LAMA above!** Only run this if you want higher quality.

In [ ]:
import os
from pathlib import Path
from demark_world.core import DeMarkWorld
from demark_world.schemas import CleanerType

# Optimize PyTorch memory management to reduce fragmentation
os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'

# Clear any cached memory
import torch
torch.cuda.empty_cache()

# Set output path
output_video_path = Path("outputs/watermark_removed_e2fgvi.mp4")

print("🚀 Processing with E2FGVI_HQ model (high quality, flicker-free)...")
print("⏱️  Typical speed: 10-20 seconds per second of video (L4 GPU)")
print("⚠️  This will take longer but produces superior results")
print("")

# Initialize and run
demark = DeMarkWorld(cleaner_type=CleanerType.E2FGVI_HQ)
demark.run(input_video_path, output_video_path)

print(f"\n✅ Done! Video saved to: {output_video_path}")
print("Proceed to Step 10 to download.")

## Step 10: Download Processed Video

In [ ]:
from google.colab import files

# Download the output video
# Use the appropriate filename based on which method you used
output_file = str(output_video_path)  # Will use the last output_video_path from Step 8 or 9

print(f"Downloading: {output_file}")
files.download(output_file)
print("✅ Download started!")

## Step 11 (Optional): Manual Batch Processing

**⚠️ Skip this if you're using Step 6.5 (Automated)**

Process multiple videos at once - upload all at once and process in sequence.

In [ ]:
# Upload multiple videos
from google.colab import files
import shutil
from pathlib import Path

print("Upload multiple video files:")
uploaded_files = files.upload()

# Create batch input/output folders
batch_input = Path("batch_input")
batch_output = Path("batch_output")
batch_input.mkdir(exist_ok=True)
batch_output.mkdir(exist_ok=True)

# Move uploaded files
for filename in uploaded_files.keys():
    shutil.move(filename, batch_input / filename)

print(f"\n✅ {len(uploaded_files)} files uploaded and ready for processing")

In [ ]:
# Process all videos in batch
from pathlib import Path
from demark_world.core import DeMarkWorld
from demark_world.schemas import CleanerType
from tqdm import tqdm

batch_input = Path("batch_input")
batch_output = Path("batch_output")

# Get all video files
video_files = list(batch_input.glob("*.mp4")) + list(batch_input.glob("*.mov")) + list(batch_input.glob("*.avi"))

print(f"Processing {len(video_files)} videos...")
print("")

# Initialize model (LAMA for speed)
demark = DeMarkWorld(cleaner_type=CleanerType.LAMA)

# Process each video
for video_file in tqdm(video_files, desc="Processing videos"):
    output_file = batch_output / f"cleaned_{video_file.name}"
    print(f"\nProcessing: {video_file.name}")

    try:
        demark.run(video_file, output_file)
        print(f"✅ Completed: {output_file.name}")
    except Exception as e:
        print(f"❌ Error processing {video_file.name}: {e}")

print("\n" + "="*60)
print("✅ Batch processing complete!")
print("="*60)

In [ ]:
# Download all processed videos
from google.colab import files
from pathlib import Path

batch_output = Path("batch_output")

print("Downloading all processed videos...")
for video_file in batch_output.glob("*.mp4"):
    print(f"Downloading: {video_file.name}")
    files.download(str(video_file))

print("\n✅ All downloads started!")

## Tips & Performance Notes

### Model Selection:
- **LAMA**: Fast (~2-5 sec/video-sec), good for quick results and testing
- **E2FGVI_HQ**: Slower (~10-20 sec/video-sec) but produces flicker-free, time-consistent results

### Supported Platforms:
DeMark-World works with watermarks from:
- ✅ OpenAI Sora (v1, v2)
- ✅ Google Gemini / Veo (3.1)
- ✅ Runway Gen-3 / Gen-4
- ✅ Pika Labs
- ✅ Kling AI
- ✅ Luma Dream Machine
- ✅ Any AI video generator with consistent watermarks

### GPU Requirements:
- **Minimum**: Google Colab T4 (free tier)
- **Recommended**: T4 or better for faster processing
- **Note**: Enable GPU in Runtime > Change runtime type

### Common Issues:
1. **Out of Memory**: Try LAMA instead of E2FGVI_HQ
2. **Slow Processing**: Ensure GPU is enabled
3. **Model Download Fails**: Check internet connection, models auto-retry

---

**Project Repository**: https://github.com/linkedlist771/DeMark-World

**License**: MIT

**Citation**:
```
@software{demarkworld2025,
  author = {linkedlist771},
  title = {DeMark-World: Universal AI Video Watermark Removal},
  year = {2025},
  url = {https://github.com/linkedlist771/DeMark-World}
}
```